_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,70 @@
+
+#### Types of responsibilities
+Instead of defining a *responsibility* in abstract terms, it may be more intuitive to list the actual types of responsibilities. Here are some examples:
+
+1. **Business logic**, for example, extracting a phone number from text, converting XML document into JSON, or classifying a money transaction as fraud. On the level of methods, such a responsibility presents itself as a single time the logic is applied, for example, extract a phone number from one given text. On the level of classes and above, it's *knowing how to do* (or encapsulating) the business function: for example, a class knowing how to convert XML documents into JSON, or a service encapsulating the detection of fraud transactions.
+2. **External integration.** On the lowest level, this can be an integration between modules within the application, such as putting a message into a `Queue` which is processed by another subsystem. Then, there are integrations with the system, such as logging or checking the system time (`System.currentTimeMillis`). Finally, there are integrations with external systems, such as database transactions, reading from or writing to a distributed message queue such as Kafka, or RPC calls to other services. On the level of methods, these responsibilities present themselves as a single transaction with the external part: for example, put a single message into a queue. On the level of classes, modules, and services, it's *knowing how to integrate* (or encapsulating integration with) the external part: for example, a class knowing how to read the system time (which is exactly what [`java.time.Clock`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/Clock.html) is), or a service encapsulating talking with an external API.
+3. **Data**: a profile of a person on a website, a JSON document, a message. Obviously, embodying a piece of data could only be a responsibility of a class (object), but not of a method, module, or service. A specific kind of data is **configuration**: a collection of parameters for some other method, class, or system.
+4. A piece of application's **control flow, execution, or data flow** model.  One example are methods that glue together calls to other methods:
+    ```
+    void processTransaction(Transaction t) {
+      if (isFraud(t)) { // Business logic
+        // External integration: logging
+        logger.info("Detected fraud transaction {}", t);
+        // Integration with external service
+        alertingService.sendAlert(new FraudTransactionAlert(t));
+      }
+    }
+    ```
+    Another example may be a `BufferedLogger` class which buffers logging statements in memory and manages a separate background thread that takes statements from the buffer and writes them to actual external logger:
+    ```
+    class BufferedLogger implements Logger {
```

Comment:
I find the `BufferedLogger` example too complex here. 30 lines of code of which only the method `writeStatementsInBackground` is really relevant. And to understand this method, you need to know what `putInterrubtibly` means and does. Can you find a simpler example (with fewer lines of code) of an Active Object that does not introduce new concepts?